dali/operators/reader/loader/webdataset_loader.h

```diff
@@ -32,8 +31,37 @@ namespace dali {
 namespace detail {
 namespace wds {
 
-const std::string kCurrentIndexVersion = "v1.2";  // NOLINT
-const std::unordered_set<std::string> kSupportedIndexVersions = {"v1.1", kCurrentIndexVersion};
+constexpr int ParseIndexVersion(const char* index_version) {
+  const char* it = nullptr;
+  for (it = index_version; *it != '\0'; it++) {}
+  int mult = 1, ret = 0;
+  auto update_ret = [&](auto it) {
+    assert(*it >= '0' && *it <= '9');
+    ret += mult * (*it - '0');
+  };
+  assert(*it == '\0');
+  update_ret(--it);
+  mult *= 10;
+  if (*--it != '.') {
+    update_ret(it--);
+  }
+  assert(*it == '.');
+  mult *= 10;
+  while (*--it != 'v') {
+    mult *= 10;
+    update_ret(it);
+  }
+  return ret;
```

Comment:
1. I think that validation and parsing could be combined.
2. This function is grossly overcomplicated.
3. `assert` is not a good way to handle invalid data - you should either throw or return an invalid value
```suggestion
  if (*it++ != 'v')
    return -1;  // or throw

  int major = 0;
  if (!isdigit(*it))
    return -1;
  for (; isdigit(*it); *it++)
    major = major * 10 + (*it - '0');
  /* you can add some sanity chceck here - e.g. that major version is less than 999 or sth */
  if (*it++ != '.')
    return -1;
  
  int minor = 0;
  if (!isdigit(*it))
    return -1;
  for (; isdigit(*it); *it++)
    minor = minor * 10 + (*it - '0');
  if (*it != 0)  // do we allow other kinds of termination/delimiters, like whitespace?
    return -1;
  /* a sanity check here to make sure that the minor version does not overflow into the major */
  if (minor > 999)
    return -1;
  return major * 1000 + minor;
```

Reply:
> you should either throw or return an invalid value

I'd go rather opposite direction - make it `nothrow`. I see two ways of using this function:
1. With hardcoded version string.
2. With user-specified version string.
In the latter case, we validate the string beforehand, so there is no way it can be incorrect. In the former case - hardcoding malformed string in the function argument will trigger the `assert`, which we use to signal coding errors.

Reply:
I've never seen a Parse function made as nothrow and a separate validation function - efficiency aside, this pattern looks foreign.
Also, if we want to check the _version_, we can make the version from numbers, not go through strings. This code is basically this:
```if (atoi(input) == atoi("2"))``` - and I hope you'll agree this is not how you compare with a number.